Trainer automatically drops unused columns in nlp datasets

sgugger opened the pull request:
Here is a basic example of use for evaluation on SST-2:
```
from nlp import load_dataset, load_metric
from transformers import AutoModelForSequenceClassification, AutoTokenizer, Trainer, TrainingArguments
dataset = load_dataset('glue', 'sst2')
metric = load_metric('glue', 'sst2')

model_name = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)

encoded_dataset = dataset.map(lambda examples: tokenizer(examples['sentence'], padding=True), batched=True)
args = TrainingArguments(output_dir = "test")

def compute_metrics(eval_pred):
    predictions, labels = eval_pred
    return metric.compute(predictions.argmax(axis=-1), labels)

trainer = Trainer(
    model,
    args,
    train_dataset=encoded_dataset["train"],
    eval_dataset=encoded_dataset["validation"],
    compute_metrics=compute_metrics,
)
trainer.evaluate()
```

The goal is to then refine this new API by trying to use it in all examples.

sgugger commented on src/transformers/trainer.py:
```diff
@@ -151,8 +162,6 @@ class Trainer:
         compute_metrics (:obj:`Callable[[EvalPrediction], Dict]`, `optional`):
             The function that will be used to compute metrics at evaluation. Must take a
             :class:`~transformers.EvalPrediction` and return a dictionary string to metric values.
-        prediction_loss_only (:obj:`bool`, `optional`, defaults to `False`):
```
Forgot to remove this in #6426

sgugger commented on src/transformers/trainer_utils.py:
```diff
@@ -53,3 +55,39 @@ class TrainOutput(NamedTuple):
 PREFIX_CHECKPOINT_DIR = "checkpoint"
+
+
+class FinalActivation(ExplicitEnum):
+    """
+    Possible values for the ``final_activation`` argument in :meth:`Trainer.from_nlp_dataset`.
+    Useful for tab-completion in an IDE.
+    """
+
+    NONE = "none"
+    ARGMAX = "argmax"
+    SIGMOID = "sigmoid"
+    SOFTMAX = "softmax"
```
Thought it was better to have some kind of enum for possible final activation functions than letting the user provide their own.

patrickvonplaten commented on src/transformers/trainer.py:
```diff
@@ -233,6 +242,83 @@ def __init__(
+            model (:class:`~transformers.PreTrainedModel`):
+                The model to train, evaluate or use for predictions.
+            args (:class:`~transformers.TrainingArguments`):
+                The arguments to tweak training.
```
(nit) "The arguments to tweak for training"

patrickvonplaten commented on src/transformers/trainer.py:
```diff
@@ -233,6 +242,83 @@ def __init__(
+        if metrics is not None:
+            compute_metrics = ComputeNLPMetrics(metrics, activation=final_activation)
+
+        # Inspect model forward signature to keep only the arguments it accepts.
```
Great functionality! Love it

patrickvonplaten commented on src/transformers/trainer.py:
```diff
@@ -233,6 +242,83 @@ def __init__(
+        cls,
+        model: PreTrainedModel,
+        args: TrainingArguments,
+        dataset: "nlp.dataset_dict.DatasetDict",
```
Think it would prefer to not have dataset as an input argument but `train_dataset` and `eval_dataset`. IMO, this has two advantages:
1) The user is not forced to use the whole dataset (e.g. if he wants to just evaluate it would be much less costly to just use the eval set for RAM) 
2) The class method signature stays the same then the init signature

also putting @lhoestq in cc here

sgugger replied:
Alternative: we can support both by having `dataset_dict`, `train_dataset` and `eval_dataset` as arguments (with an assert to check that not both are provided at the same type).

patrickvonplaten replied:
Is there a big advantage to having `dataset_dict` as an input? It saves a couple of lines of code, but I don't really see a big argument for it

sgugger replied:
The most common use case of Trainer is train + eval, and the preprocessing is often the same for both training and evaluation, so IMO it's more useful to support `dataset_dict`.

If we go down the road of "it just saves a few lines of code", then we can just say there is no need for a class method and the user can use the regular init ;-)

thomwolf replied:
I think I would go with the more flexible `train_dataset` and `eval_dataset`, in particular since the internal names for these in a dict might be different (in particular `test` can sometimes be `validation`).

It doesn't even save you a line of code in most cases (you just split the dict in the call) so better to have it explicit imo.

lhoestq replied:
I agree that from the user's perspective it can be confusing to see that the input expects a dataset dict. Asking explicitly for the train/val/test datasets splits is clearer imo

patrickvonplaten commented on src/transformers/trainer.py:
```diff
@@ -233,6 +242,83 @@ def __init__(
+        """
+        assert is_nlp_available(), "This method requires the nlp library: `pip install nlp`."
+        if metrics is not None:
+            compute_metrics = ComputeNLPMetrics(metrics, activation=final_activation)
```
Intuitively I would say that the metric is responsible to choose which "final_activation" to use here, so I would move this into the metrics instead and think the user should not even have to specific it. Or is it very common that the same metric has multiple possible "final_activation" functions that can be applied?

sgugger replied:
`nlp` metrics don't have an activation built in. IMO a metric is just a function, it should not have an idea of what the final activation is.
It's a task model that has an idea of the activation function, though it would not know if it needs to take argmax (for metrics that uses prediction indices like accuracy) or softmax (for metrics that use the probabilities like roc_score, auc_score). In the end, I think we could have a default that works most of the time by mapping models to a given final activation, but it still needs to be available as an argument to the user for all the cases that can go wrong.

LysandreJik commented on src/transformers/trainer.py:
```diff
@@ -233,6 +242,83 @@ def __init__(
+        signature = inspect.signature(model.forward)
+        signature_columns = list(signature.parameters.keys())
+        # Labels may be named label or label_ids, the default data collator handles that.
+        signature_columns += ["label", "label_ids"]
+        dataset_columns = dataset[list(dataset.keys())[0]].column_names
+        columns = [k for k in signature_columns if k in dataset_columns]
+        dataset.set_format(columns=columns)
```
Could we also print a `logger.warning` if some values are skipped?

sgugger replied:
Did an info instead of a warning since it's what the API is supposed to do and users are scared of warnings. Also properly documented the API is doing that.

sshleifer commented:
This is sweet!

thomwolf commented on src/transformers/trainer.py:
```diff
@@ -233,6 +243,117 @@ def __init__(
+        model: PreTrainedModel,
+        args: TrainingArguments,
+        dataset_dict: "nlp.dataset_dict.DatasetDict" = None,
+        train_dataset: "nlp.dataset_dict.Dataset" = None,
+        eval_dataset: "nlp.dataset_dict.Dataset" = None,
```
probably more `nlp.Dataset`

thomwolf commented on src/transformers/trainer.py:
```diff
@@ -233,6 +243,117 @@ def __init__(
+        eval_dataset: "nlp.dataset_dict.Dataset" = None,
+        data_collator: Optional[DataCollator] = None,
+        metrics: Optional[Union["nlp.Metric", List["nlp.Metric"]]] = None,
+        final_activation: Optional[Union[str, FinalActivation, Callable]] = None,
```
I think I would call this `metrics_preprocessing` and accept either a callable or an activation string/enum

thomwolf reviewed (commented):
Overall and from a user-API point of view I'm actually wondering if we really need a new `classmethod` specifically designed for `nlp` here.

In particular, given that we've built `nlp` with the goal of being very explicit about everything and having a dataset that behaves like a normal python container.

As I see it there are a few elements added by this method:
- filtering the inputs of the model but this is already done by default in the tokenizers themselves (to avoid adding unnecessary inputs) and maybe could be just improved by testing the format of the dataset if the dataset is an instance of `nlp.Dataset` in the __init__ method and setting the format if some columns cannot be handled by the model. Be aware that the format is a stateful property of the dataset so this actually has side effects (maybe we could change this).
- handling metrics but maybe this could be obtained by working on the `nlp ` side to add support for model outputs and on this side by allowing an `nlp.Metric` to be passed as `compute_metrics` in the normal __init__ method.

Let's have a quick discussion about this maybe.

sgugger commented:
Removed all the changes linked to metrics and moved the column dropping to anywhere we pass a Dataset (init, evaluate and predict). As discussed, we'll propose an API for the metrics once we have changed all examples to use `Trainer` and `nlp`, so we know exactly what the API has to support.
